pymobiledevice3/remote/remotexpc.py

```diff
@@ -0,0 +1,331 @@
+            'data': payload,
+        }
+    elif type(payload).__name__ == 'XpcUInt64Type':
+        xpc_object = {
+            'type': XpcMessageType.UINT64,
+            'data': payload,
+        }
+    elif type(payload).__name__ == 'XpcInt64Type':
+        xpc_object = {
+            'type': XpcMessageType.INT64,
+            'data': payload,
+        }
+    else:
+        raise Exception(f'unrecognized type for: {payload} {type(payload)}')
+
+    return xpc_object
+
+
+def create_xpc_wrapper(d: Mapping, message_id: int = 0) -> bytes:
+    flags = XpcFlags.ALWAYS_SET
+    if len(d.keys()) > 0:
+        flags |= XpcFlags.DATA_PRESENT
+
+    xpc_payload = {
+        'message': build_xpc_object(d)
+    }
+
+    xpc_wrapper = {
+        'flags': flags,
+        'size': len(XpcPayload.build(xpc_payload)),
+        'message_id': message_id,
+        'payload': xpc_payload,
+    }
+    return XpcWrapper.build(xpc_wrapper)
+
+
+H2FRAME_SIZE = len(H2Frame())
+HTTP2_MAGIC = b'PRI * HTTP/2.0\r\n\r\nSM\r\n\r\n'
+
+
+class RemoteXPCConnection:
+    def __init__(self, address: Tuple[str, int]):
+        self.address = address
+        self.sock: Optional[socket.socket] = None
+        self.next_message_id = 0
+        self.peer_info = None
+
+    def __enter__(self) -> 'RemoteXPCConnection':
+        self.connect()
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
+        self.close()
+
+    def connect(self) -> None:
+        self.sock = create_connection(self.address)
+        self._do_handshake()
+
+    def close(self) -> None:
+        self.sock.close()
+
+    def send_request(self, data: Mapping) -> None:
+        packet = H2Frame(type=0, stream_id=1) / H2DataFrame(
+            data=create_xpc_wrapper(data, message_id=self.next_message_id))
+        self.sock.sendall(bytes(packet))
+        # self.next_message_id += 1
+
+    def receive_response(self):
+        while True:
+            frame = self._receive_h2_frame()
+            if isinstance(frame.lastlayer(), H2ResetFrame):
+                raise StreamClosedError()
+            if not isinstance(frame.lastlayer(), H2DataFrame):
+                continue
+            xpc_wrapper = XpcWrapper.parse(frame.data)
+            if xpc_wrapper.payload is None:
+                continue
+            if xpc_wrapper.payload.message.data.entries is None:
+                continue
+
+            self.next_message_id = xpc_wrapper.message_id + 1
+            return get_object_from_xpc_wrapper(frame.data)
+
+    def send_receive_request(self, data: Mapping):
+        self.send_request(data)
+        return self.receive_response()
+
+    def _do_handshake(self) -> None:
+        self.sock.sendall(HTTP2_MAGIC)
+
+        settings_frame = H2Frame() / H2SettingsFrame()
+        settings_frame.settings = [
+            H2Setting(id=H2Setting.SETTINGS_MAX_CONCURRENT_STREAMS, value=100),
+            H2Setting(id=H2Setting.SETTINGS_INITIAL_WINDOW_SIZE, value=1048576),
```

Comment:
Move to const

Reply:
Also, it prettier in hex

Reply:
that's how it appears in wireshark. so i'll make it const and explain that is the representation to make it easier to later validate

Reply:
refactored